src/github.com/getlantern/errors/errors.go

```diff
@@ -0,0 +1,641 @@
+/*
+Package errors defines error types used across Lantern project.
+
+  errors.ReportTo(myErrorReporter)
+  ...
+  if n, err := Foo(); err != nil {
+    errors.Wrap(err).Report()
+  }
+
+Wrap() method will try as much as possible to extract details from the error
+passed in, if it's errors defined in Go standard library. For application
+defined error type, at least the Go type name and what err.Error() returns
+will be recorded.
+
+Extra fields can be chained in any order, at any time.
+
+  func Foo() *Error {
+  	//...
+    return errors.New("some error").With("some_counter", 1).WithOp("connect")
```

Comment:
I choose `WithOp` simply because `Op` is already a struct field, other method calls are without "With". Nothing significant, but I choose to lazily create ProxyingInfo/UserLocale/UserAgentInfo etc. Method call can hide the details of how the fields are structured. Only the reporter needs those details.

If we allow manually creating Error, then the struct should be flat, or we'll end up with:
```
return &errors.Error{
    Desc: "some error",
    Op: "connect",
    ProxyingInfo: &errors.ProxyingInfo{
      ProxyType: errors.NoProxy,
      ProxyAddr: "...",
    }
    Extra: map[string]string{
        "some_counter": 1,
    }
}
```
instead of
```
return errors.New("some error").WithOp("connect").
    ProxyType(errors.NoProxy).ProxyAddr("...").With("some_counter", 1)
```